crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,136 @@
+/// will be preserved. Note that this function will always remove all trailing zeros
+/// for microseconds.
+pub fn time_starting_with_ms_with_precision(
+    duration: &Duration,
+    precision: u8,
+) -> String {
+    let (d, h, m, s, ms, mcs) = get_duration_parts(duration);
+    match (d, h, m, s) {
+        (0, 0, 0, 0) => {
+            let ms_mcs = ms as f64 + mcs as f64 / 1000.0;
+            format!("{} ms", round_with_precision(ms_mcs, precision))
+        }
+        (d, h, m, s) => format_dhms(d, h, m, s),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn duration_milli_micro(milliseconds: u16, microseconds: u16) -> Duration {
+        let microseconds = microseconds as u64;
+        let milliseconds = 1000 * milliseconds as u64;
+        Duration::from_micros(milliseconds + microseconds)
+    }
+
+    #[test]
+    fn test_time_starting_with_seconds() {
+        let f = |duration| time_starting_with_seconds(&duration);
+        fn duration(days: u16, hours: u8, minutes: u8, seconds: u8) -> Duration {
+            let seconds = seconds as u64;
+            let minutes = 60 * minutes as u64;
+            let hours = 60 * 60 * hours as u64;
+            let days = 24 * 60 * 60 * days as u64;
+            Duration::from_secs(days + hours + minutes + seconds)
+        }
+        assert_eq!(" 0 s", &f(duration(0, 0, 0, 0)));
+        assert_eq!("59 s", &f(duration(0, 0, 0, 59)));
+        assert_eq!(" 1 m 12 s", &f(duration(0, 0, 1, 12)));
+        assert_eq!("59 m  0 s", &f(duration(0, 0, 59, 0)));
+        assert_eq!(" 5 h  1 m  2 s", &f(duration(0, 5, 1, 2)));
+        assert_eq!("  1 d  0 h  0 m  0 s", &f(duration(1, 0, 0, 0)));
+        assert_eq!(" 69 d  0 h  0 m  0 s", &f(duration(69, 0, 0, 0)));
+        assert_eq!("100 d 10 h 10 m 10 s", &f(duration(100, 10, 10, 10)));
+    }
+
+    #[test]
+    fn test_time_as_ms_with_precision_1() {
+        let f = |duration| time_starting_with_ms_with_precision(&duration, 1);
+        let duration = duration_milli_micro;
+        assert_eq!("123.5 ms", &f(duration(123, 456)));
+        assert_eq!("123.5 ms", &f(duration(123, 455)));
+        assert_eq!("123.4 ms", &f(duration(123, 445)));
+        assert_eq!("123.4 ms", &f(duration(123, 440)));
+        assert_eq!("123.1 ms", &f(duration(123, 100)));
+        assert_eq!("123 ms", &f(duration(123, 0)));
+    }
+
+    #[test]
+    fn test_time_as_ms_with_precision_2() {
+        let f = |duration| time_starting_with_ms_with_precision(&duration, 2);
+        let duration = duration_milli_micro;
+        assert_eq!("123.46 ms", &f(duration(123, 456)));
+        assert_eq!("123.46 ms", &f(duration(123, 455)));
+        assert_eq!("123.45 ms", &f(duration(123, 454)));
+        assert_eq!("123.45 ms", &f(duration(123, 450)));
+        assert_eq!("123.1 ms", &f(duration(123, 100)));
+        assert_eq!("123 ms", &f(duration(123, 0)));
+        // assert_eq!(format!("{:?}", &f(duration(123, 0))).as_str(), &f(duration(123, 0)));
```

Comment:
why is this commented?

Reply:
I think I wanted to make another commit with the last change, probably regarding [this](https://github.com/typst/typst/pull/4587#issuecomment-2291389171). But currently yeah, it's not removed, and I remembered that after latest commit. I guess this can be counted as an oopsie.